### ISSUE TITLE
Implement substitutions for smarter docs 

### ISSUE THREAD

**Issue by MatthewCroughan** [foundriesio/docs#164]
I want to put simple variables/substitutions into the docs. This has proven hard, because Sphinx does not have a pre-processor with which to replace text, it uses RST and inherits basic substitution as a result

Assuming the substitution `|ARCH|` is defined as `x86_64` by having `.. |ARCH| replace:: x86_64` set in an included file:

```
├── reference-manual
│   ├── qemu
│   │   ├── arm64.rst
│   │   ├── arm64-substitutions.yaml <---- 
│   │   ├── generic-prepare.template
│   │   ├── qemu.rst
│   │   ├── riscv64.rst
│   │   ├── riscv64-substitutions.yaml <---- 
│   │   ├── x86_64.rst
│   │   └── x86_64-substitutions.yaml <---- 
```

We cannot do the following:

### Conditionally/Variably include documents based ARCH.

`.. include:: |ARCH|-qemu-command.rst`

`.. asciinema:: ./demo/|ARCH|.cast`

### We can do things like this, though:

```
.. |ARCH| replace:: x86_64
.. |MACHINE| replace:: intel-corei7-64
.. |FIRMWARE_BLOB| replace:: ovmf.qcow2

.. parsed-literal::

     └── |ARCH|
         ├── lmp-factory-image-|MACHINE|.wic.gz
         └── |FIRMWARE_BLOB|
```

Resolves to:

```
     └── x86_64
         ├── lmp-factory-image-intel-corei7-64.wic.gz
         └── ovmf.qcow2
```

**Comment by MatthewCroughan** [comment]
I got some great help from the author of the sphinx-ext-substitution plugin, which basically does exactly what is needed.
https://github.com/NordicHPC/sphinx_ext_substitution/issues/1#issuecomment-767655825

**Comment by MatthewCroughan** [comment]
This extension works as follows:

```
 `:sub`(VARIABLE_NAME)`
```

```
**Necessary artifacts**

.. parsed-literal::

     └── :sub:`(ARCH)`
         ├── lmp-factory-image-:sub:`(MACHINE)`.wic.gz
         └── :sub:`(FIRMWARE_BLOB)`
```
Resolves to:
```
└── x86_64
    ├── lmp-factory-image-intel-corei7-64.wic.gz
    └── ovmf.qcow2
```

**Comment by MatthewCroughan** [comment]
I've just discovered that I can't do what I want with the `sphinx_ext_substitution` extension, since I cannot conditionally include variables like this:

ARCH is defined separately in each `*-substitutions` file, it is not specified at build time:

`riscv64-substitutions.inc`
```
.. |ARCH| replace:: riscv64
.. |MACHINE| replace:: qemuriscv64
.. |FIRMWARE_BLOB| replace:: fw_payload.elf
```

Then, `generic-template.template` contains no `.. include`

`generic-template.template`

```
**Necessary artifacts**

.. parsed-literal::

     └── |ARCH|
         ├── lmp-factory-image-|MACHINE|.wic.gz
         └── |FIRMWARE_BLOB|
```

then the files pertaining to arch like `riscv64.rst` `arm64.rst`, etc, include the `generic-template.template` and the overrides/substitutions file.

`riscv64.rst`
```
riscv64
=======

.. include:: generic-template.template
.. include:: riscv64-substitutions.inc
```

This allows the `.template` to be filled out, and each file gets a different set of substitutions.  

# TL;DR

I believe the solution is templating using Jinja as a pre-processor. This blog post from the founder of readthedocs has more info on that. https://www.ericholscher.com/blog/2016/jul/25/integrating-jinja-rst-sphinx/

**Comment by MatthewCroughan** [comment]
Yocto also uses Sphinx. They solve this by writing their own pre-processor in Python:

http://git.yoctoproject.org/cgit/cgit.cgi/yocto-docs/commit/documentation/sphinx/yocto-vars.py?id=dc5f53fae8fdfdda04285869dd1419107b920bfe

https://git.yoctoproject.org/cgit/cgit.cgi/yocto-docs/commit/documentation/sphinx/yocto-vars.py?id=59537c7fa49e3ea6918f45b3201ad16d56988b9b

I don't like the idea of any of what I've proposed so far. But MkDocs seems to have all this functionality built in, and exposes Jinja to the user by default, in a way that makes more sense. 
https://squidfunk.github.io/mkdocs-material/reference/variables/

**Comment by MatthewCroughan** [comment]
MkDocs provides rich variable manipulation using Jinja, in a way that Sphinx fails to without custom code.

https://github.com/fralau/mkdocs_macros_plugin#defining-variables
> ### Defining variables
> Regular **variables** can be defined in five ways:
> 
> | No | Validity | For whom | Description |
> | --- | --- | --- | ---- |
> | 1. | global | designer of the website | in the `mkdocs.yml` file, under the `extra` heading |
> | 2. | global | contributor | in external yaml definition files |
> | 3. | global | programmer | in a `main.py` file (Python), by adding them to a dictionary |
> | 4. | local (page) | writer | in the YAML header of each Markdown page |
> | 5. | local (page) | writer | with a `{%set variable = value %}`  statement |

**Comment by MatthewCroughan** [comment]
It is unlikely that we will be able to implement any of this with Sphinx, even if we could it's my opinion that it would be too difficult to use from the writer's perspective. 

I'm going to spin out a new issue that suggests moving from Sphinx to MkDocs so we can make use of this functionality.

**Comment by kprosise** [comment]
closing issue, added as comment supporting change away from Sphinx on confluence page, as well as opened an issue on jira
